libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/navigation/NavigationRouteProcessorTest.java

```diff
@@ -0,0 +1,184 @@
+
+import static junit.framework.Assert.assertEquals;
+import static junit.framework.Assert.assertNotNull;
+import static junit.framework.Assert.assertTrue;
+import static org.mockito.Mockito.mock;
+
+@RunWith(RobolectricTestRunner.class)
+@Config(constants = BuildConfig.class, manifest = Config.DEFAULT_MANIFEST_NAME)
+public class NavigationRouteProcessorTest extends BaseTest {
+
+  private NavigationRouteProcessor routeProcessor;
+  private MapboxNavigation navigation;
+
+  @Before
+  public void setup() throws Exception {
+    routeProcessor = new NavigationRouteProcessor();
+    MapboxNavigationOptions options = MapboxNavigationOptions.builder().build();
+    navigation = new MapboxNavigation(mock(Context.class), ACCESS_TOKEN, options, mock(NavigationTelemetry.class),
+      mock(LocationEngine.class));
+    navigation.startNavigation(buildDirectionsRoute());
+  }
+
+  @Test
+  public void sanity() throws Exception {
+    assertNotNull(routeProcessor);
+  }
+
+  @Test
+  public void onFirstRouteProgressBuilt_newRouteIsDecoded() throws Exception {
+    RouteProgress progress = routeProcessor.buildNewRouteProgress(navigation, mock(Location.class));
+    assertEquals(0, progress.legIndex());
+    assertEquals(0, progress.currentLegProgress().stepIndex());
+  }
+
+  @Test
+  public void onShouldIncreaseStepIndex_indexIsIncreased() throws Exception {
+    RouteProgress progress = routeProcessor.buildNewRouteProgress(navigation, mock(Location.class));
+    int currentStepIndex = progress.currentLegProgress().stepIndex();
+    routeProcessor.onShouldIncreaseIndex();
+    routeProcessor.checkIncreaseStepIndex(navigation);
+    RouteProgress secondProgress = routeProcessor.buildNewRouteProgress(navigation, mock(Location.class));
+    int secondStepIndex = secondProgress.currentLegProgress().stepIndex();
+    assertTrue(currentStepIndex != secondStepIndex);
+  }
+
+  @Test
+  public void onSnapToRouteEnabledAndUserOnRoute_snappedLocationReturns() throws Exception {
+    // Decode step points for snap-to-route logic
+    RouteProgress progress = routeProcessor.buildNewRouteProgress(navigation, mock(Location.class));
+
+    boolean snapEnabled = true;
+    boolean userOffRoute = false;
+
+    LegStep currentStep = progress.currentLegProgress().currentStep();
+
+    LineString lineString = LineString.fromPolyline(currentStep.geometry(), Constants.PRECISION_6);
+    List<Point> coordinates = lineString.coordinates();
+
+    Point lastPointInCurrentStep = coordinates.remove(coordinates.size() - 1);
+    Location rawLocation = buildDefaultLocationUpdate(
+      lastPointInCurrentStep.longitude(), lastPointInCurrentStep.latitude()
+    );
+
+    Location snappedLocation = routeProcessor.buildSnappedLocation(
+      navigation, snapEnabled, rawLocation, progress, userOffRoute
+    );
+
+    assertTrue(!rawLocation.equals(snappedLocation));
+  }
+
+  @Test
+  public void onSnapToRouteDisabledAndUserOnRoute_rawLocationReturns() throws Exception {
```

Comment:
Normally is good to structure the tests following the triple A 👀 https://speakerdeck.com/guardiola31337/elegant-unit-testing-mobilization-2016?slide=40 http://wiki.c2.com/?ArrangeActAssert
If the `A`rrange part gets too long we should create factory or builder methods.

This applies to other tests included in `NavigationRouteProcessorTest`.